test/integration/replicaset/replicaset_test.go

```diff
@@ -515,3 +519,682 @@ func TestRSSelectorImmutability(t *testing.T) {
+	}
+	//Verify all pods are not available manually
+	podSelector := labels.Set{"foo": "bar"}.AsSelector()
+	options := metav1.ListOptions{LabelSelector: podSelector.String()}
+
+	// Make .spec.minReadySeconds = 0 so that we can test the replicas are available
+	rs.Spec.MinReadySeconds = 0
+	rs, err = c.ExtensionsV1beta1().ReplicaSets(ns.Name).Get(rs.Name, metav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("failed to get replicaset %s: %v", rs.Name, err)
+	}
+	rs, err = c.ExtensionsV1beta1().ReplicaSets(ns.Name).Update(rs)
+	if err != nil {
+		t.Fatalf("failed to update replicaset %s: %v", rs.Name, err)
+	}
+	waitRSStable(t, c, rs, ns.Name)
+
+	// After updating the RS and waiting RS to be stable, we still have to GET the RS
+	// for the RS to reflect newest .Status
+	rs, err = c.ExtensionsV1beta1().ReplicaSets(ns.Name).Get(rs.Name, metav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("failed to get replicaset %s: %v", rs.Name, err)
+	}
+	if rs.Status.Replicas != 3 {
+		t.Fatalf("Unexpected .Status.Replicas: Expected 3, saw %d\n", rs.Status.Replicas)
+	}
+	if rs.Status.ReadyReplicas != 3 {
+		t.Fatalf("Unexpected .Status.ReadyReplicas: Expected 3, saw %d\n", rs.Status.ReadyReplicas)
+	}
+	// Verify all pods are available manually
+	pods, err := c.Core().Pods(ns.Name).List(options)
+	if err != nil {
+		t.Fatalf("Failed obtaining a list of pods that match the pod labels %v: %v", labelMap, err)
+	}
+	if pods == nil {
+		t.Fatalf("Obtained a nil list of pods")
+	}
+	for _, pod := range pods.Items {
+		if podutil.IsPodAvailable(&pod, rs.Spec.MinReadySeconds, metav1.Now()) {
+			t.Fatalf("Pod %s is not available after .spec.minReadySeconds of RS %s changed to %d", pod.Name, rs.Name, rs.Spec.MinReadySeconds)
+		}
+	}
+}
+
+func TestReplicaSetBasics(t *testing.T) {
+	s, closeFn, rm, informers, c := rmSetup(t)
+	defer closeFn()
+	ns := framework.CreateTestingNamespace("test-replicaset-basics", s, t)
+	defer framework.DeleteTestingNamespace(ns, s, t)
+	labelMap := map[string]string{"foo": "bar"}
+	rs := newRS("rs", ns.Name, 2)
+	rs.Spec.Selector.MatchLabels = labelMap
+	rs.Spec.Template.Labels = labelMap
+
+	// Create 2 pods (one is failed, one is being deleted)
+	// Ensure both are being created again in the end
+	failedPod := newMatchingPod("failed-pod", ns.Name)
+	failedPod.Labels = labelMap
+	failedPod.Status.Phase = v1.PodFailed
+	deletedPod := newMatchingPod("deleted-pod", ns.Name)
+	deletedPod.Labels = labelMap
+	deletedPod.DeletionTimestamp = &metav1.Time{Time: time.Now()}
+	createRSsPods(t, c, []*v1beta1.ReplicaSet{rs}, []*v1.Pod{failedPod, deletedPod}, ns.Name)
+
+	stopCh := make(chan struct{})
+	informers.Start(stopCh)
+	go rm.Run(5, stopCh)
+	defer close(stopCh)
+	waitRSStable(t, c, rs, ns.Name)
+
+	rs, err := c.ExtensionsV1beta1().ReplicaSets(ns.Name).Get(rs.Name, metav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("failed to get replicaset %s: %v", rs.Name, err)
+	}
+	if rs.Status.Replicas != 2 {
```

Comment:
waitRSStable already ensured Status.Replicas is what we expect. Same below.

Reply:
Ya, there are many redundant/unnecessary `waitRSStable` calls.